BitList.cs

```diff
@@ -0,0 +1,1552 @@
+            return true;
+        }
+    }
+
+    /// <summary>
+    /// Insert the specified list of bit in the specified index.
+    /// </summary>
+    /// <exception cref="IndexOutOfRangeException" />
+    public void Insert(int index, BitList bits)
+    {
+        if (IsEmpty)
+            this.bits = new BitArray(bits.bits);
+        else if (index < bits.Count && index >= 0)
+        {
+            var bools = ToBooleanArray() as IEnumerable<bool> as List<bool>;
+            bools.InsertRange(index, bits.ToBooleanArray());
+            this.bits = new BitArray(bools.ToArray());
+        }
+        else
+            throw new IndexOutOfRangeException();
+    }
+
+    /// <summary>
+    /// Insert the specified bit in the specified index.
+    /// </summary>
+    /// <exception cref="IndexOutOfRangeException" />
+    public void Insert(int index, bool bit)
+    {
+        if (IsEmpty)
+            bits = new BitArray(new[] { bit });
+        else if (index < bits.Length && index >= 0)
+        {
+            var bools = ToBooleanArray() as IEnumerable<bool> as List<bool>;
+            bools.Insert(index, bit);
+            bits = new BitArray(bools.ToArray());
+        }
+        else
+            throw new IndexOutOfRangeException();
+    }
+
+    /// <summary>
+    /// Replace a list of bit of the current instance with another list of bit.
+    /// </summary>
+    /// <exception cref="EmptyInstanceException" />
+    public void Replace(BitList oldBitList, BitList newBitList)
+    {
+        if (IsEmpty)
+            throw new EmptyInstanceException();
+        else
+        {
+            var thisBitListText = ToText();
+            var oldBitListText = oldBitList.ToText();
+            var newBitListText = newBitList.ToText();
+            var bitsText = thisBitListText.Replace(oldBitListText, newBitListText);
+            bits = Parse(bitsText).bits;
+        }
+    }
+
+    /// <summary>
+    /// Set the value in the specified index with a new value.
+    /// </summary>
+    /// <exception cref="ArgumentOutOfRangeException" />
+    public void Set(int index, bool value)
+    {
+        if (!IsEmpty && (index < bits.Length | index >= 0))
+            bits[index] = value;
+        else
+            throw new ArgumentOutOfRangeException();
+    }
+
+    /// <summary>
+    /// Set the value in the specified index with a new collection of values.
+    /// </summary>
+    /// <exception cref="IndexOutOfRangeException" />
+    public void Set(int index, BitList value)
+    {
+        if (!IsEmpty && index < bits.Length && index >= 0 && value.Count <= bits.Length - index && value.Count > 0)
+        {
+            var bools = ToBooleanArray() as IEnumerable<bool> as List<bool>;
+            bools.RemoveRange(index, value.Count);
+            bools.InsertRange(index, value.ToBooleanArray());
+            bits = new BitArray(bools.ToArray());
+        }
+        else
+            throw new IndexOutOfRangeException();
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: ['System.IndexOutOfRangeException' should not be thrown by user code.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=5545753)